### ISSUE TITLE
Integrate in blender 2.79

### ISSUE THREAD

**Issue by PavelBlend** [doyubkim/fluid-engine-dev#191]

**Comment by doyubkim** [comment]
Thanks for the suggestion. Integrating Jet Framework into 3D packages such as Blender3D, Maya, Houdini, etc would be amazing. In fact, Kent Barber ported Jet into a Cinema4D plugin (http://www.jetfluids.com). But since I do now have that much of experience with writing Blender3D add-ons and my bandwidth is quite full already, I'd like to keep this as an open issue to everyone who wants to contribute.

**Comment by sachk** [comment]
It'd be better to implement it in 2.8, since there's so many changes and it's coming out fairly soon. Blender's API is fairly easy to use, you could take a look at this plugin for an example.

https://github.com/rlguy/Blender-FLIP-Fluids

**Comment by doyubkim** [comment]
Thanks, @sachk for the reference!

**Comment by doyubkim** [comment]
Hi Pavel,

The resolution here is the simulation resolution. To increase the number of particles, you should change the spacing parameter for VolumeParticleEmitter3.

**Comment by doyubkim** [comment]
Btw, it’s super impressive to see Jet Framework working with Blender! Let me know if you encounter any other issues. Once it becomes mature, it would be great to try it out from my end as well. Anyway, awesome stuff!

**Comment by PavelBlend** [comment]
> Let me know if you encounter any other issues.

So far I have one problem: poor performance. We have to wait a long time for the calculation of fluid. Maybe I'm not setting the parameters correctly. Or the simulator initially works so slowly?
The gif animation above is a simulation that took 10 minutes.

**Comment by doyubkim** [comment]
Can you share your code? Would like to reproduce from my end and see what’s causing the issue. 
Also, is this based on the release or debug build? What is your system configuration? (OS, memory, CPU)

**Comment by PavelBlend** [comment]
> Also, is this based on the release or debug build?

debug build. I could not compile the Release version.

> What is your system configuration? (OS, memory, CPU)

os: windows 7 64 bit
memory: 8 GB
CPU: AMD A8 5600K 3.60 GHz 4 Core

**Comment by PavelBlend** [comment]
Code:
https://drive.google.com/open?id=1fP0QJTFT2LMr9n-f5FmW4aUUouqF2pH7

Works in a daily assembly 2.79:
https://builder.blender.org/download/

**Comment by PavelBlend** [comment]
Performance problem partially solved. The problem was the same on my part.
But still the performance is low. At 100 resolution baked for a long time.

**Comment by PavelBlend** [comment]
How to build a mesh of fluid particles using pyjet?

**Comment by doyubkim** [comment]
I looked into your code and I think I have a couple of things I could try out (for both performance and mesh export). I think Python API was not very clear about it and also missing some features compared to C++ API. Let me post update soon.

**Comment by doyubkim** [comment]
Ok, so a new PR is posted. Also added a Python example (`apic_example03.py`) to show your use case such as using `TriangleMesh3` for emitter (using `ImplicitTriangleMesh3` for accurate/fast particle generation) and exporting particles as a surface mesh (`marchingCubes`). It also shows a new trick to make sim a bit faster if water is not occupying entire domain (`solver.useCompressedLinearSystem = True`). This example took me a couple of seconds per frame.

**Comment by PavelBlend** [comment]
All my code is now here:
https://github.com/PavelBlend/blender_jet_fluids_addon

I have a question: why does Bunny have extra particles near the ears (see the gif animation from the link above)? I used the standard bunny object from your repository. Is the particle generator not working correctly?

**Comment by doyubkim** [comment]
You should use ImplicitTriangleMesh3 as a converter before assigining a mesh to the emitter. (Please check out the example I mentioned above.) This is indeed a bit confusing and will be fixed in the future. For now, that’s how it should work.

**Comment by doyubkim** [comment]
Can you provide the mesh and the resolution setting you are using? Seems like the SDF of the collider is messed up.

**Comment by PavelBlend** [comment]
[test_emiiter_particles.zip](https://github.com/doyubkim/fluid-engine-dev/files/2715423/test_emiiter_particles.zip)
This effect appears at any resolution.

**Comment by doyubkim** [comment]
Just checked out the meshes, but `collider.obj` and `emitter.obj` are identical.

**Comment by PavelBlend** [comment]
[meshes_test.zip](https://github.com/doyubkim/fluid-engine-dev/files/2717069/meshes_test.zip)

**Comment by PavelBlend** [comment]
Question:
Is it possible to compile pyjet for python 3.5 without changing the source code?
I need pyjet.pyd for python 3.5 since blender 2.79 uses exactly this version of python.
But I also need pyjet for python 3.7.
Since this version of python is used in daily builds and in version 2.80.
And can you compile and release pyjet for Linux and MacOS (python 3.5 and python 3.7)?

**Comment by doyubkim** [comment]
Hi @PavelBlend !
I often use Miniconda Python distribution to build and test different versions of Python. If you need pyjet targeting Python 3.5, you can try creating a conda environment with Python 3.5 (something like `conda create -n py35 python=3.5`), activate that environment (`source activate py34`) and build pyjet (`pip install -U .` from the repository's root directory).
I am thinking of distributing pyjet's binary wheel at some point (no timeline though), but the above method should unblock you in the meantime.

**Comment by PavelBlend** [comment]
Question:
Why is the pyjet.ImplicitTriangleMesh3 class being created for a long time?
Maybe I do not correctly specify the parameters?
I create a solver:
`solver = pyjet.PicSolver3(resolution=(100, 100, 100), gridOrigin=(0, 0, 0), domainSizeX=1)`

How do I create pyjet.ImplicitTriangleMesh3? What should be resolutionX? 100?:
`imp_triangle_mesh = pyjet.ImplicitTriangleMesh3(mesh=triangle_mesh, resolutionX=100, margin=0.2)`

Or you need to take into account the dimensions of the 3D model?:
![01](https://user-images.githubusercontent.com/7983249/54761408-186e3780-4c03-11e9-992b-b15cf4ab88cc.jpg)

**Comment by doyubkim** [comment]
ImplicitTriangleMesh3.resolutionX should be high enough to capture geometric details of the mesh. For your example, I would use 100 * mesh_size / solver_size.

And to be honest, ImplicitTriangleMesh3 has been bugging me a lot, too (API is weird, perf is slow, etc). I am actually implementing an alternative approach which is still under tests. Let me try to create a shared branch so we can further test on it.

**Comment by PavelBlend** [comment]
I have released a new version of the jet-fluids addon for blender 2.81:
https://github.com/PavelBlend/blender_jet_fluids_addon/releases
There is no documentation for the addon yet.

**Comment by doyubkim** [comment]
Amazing work, @PavelBlend !!
